### PR TITLE
Example dag slackfile

### LINKED ISSUE #17368: Add example DAG for SlackAPIFileOperator
The SlackAPIFileOperator is not straight forward and it might be better to add an example DAG to demonstrate the usage.

### PR DESCRIPTION
**Pull request by subkanthi**
closes: #17368 
- Fixed bug with SlackAPIFileOperator with uploading file. 
- Added example dag to demonstrate support for uploading file and file content.

### CODE REVIEW

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
some docs and test failures

File: airflow/utils/log/file_task_handler.py [hunk 1]
```diff
@@ -187,10 +187,14 @@ def _read(self, ti, try_number, metadata=None):
...
-                    log += "*** See more at https://airflow.apache.org/docs/apache-airflow/" \
-                           "stable/configurations-ref.html#secret-key\n***"
+                    log += (
+                        "*** !!!! Please make sure that all your webservers and workers have"
+                        " the same 'secret_key' configured in 'webserver' section !!!!!\n***"
+                    )
+                    log += (
+                        "*** See more at https://airflow.apache.org/docs/apache-airflow/"
+                        "stable/configurations-ref.html#secret-key\n***"
+                    )
```
Comment (eladkal) [inline airflow/utils/log/file_task_handler.py]: This isn't related to this PR

Comment (potiuk) [reply airflow/utils/log/file_task_handler.py]: I broke it and just fixed it in main (bad Jarek!)

Comment (potiuk) [reply airflow/utils/log/file_task_handler.py]: Needs rebase

File: airflow/providers/slack/operators/slack.py [hunk 2]
```diff
@@ -195,49 +196,50 @@ class SlackAPIFileOperator(SlackAPIOperator):
...
-    template_fields = ('channel', 'initial_comment', 'filename', 'filetype', 'content')
+    template_fields = ('channel', 'initial_comment', 'filetype', 'content')
     ui_color = '#44BEDF'
 
     def __init__(
         self,
         channel: str = '#general',
         initial_comment: str = 'No message has been set!',
-        filename: str = None,
+        file: str = None,
```
Comment (mik-laj) [inline airflow/providers/slack/operators/slack.py]: It looks like a breaking change. Is there any way we can keep backward compatibility? If not, can you add a note to the changelog? For example note, see: google provider https://github.com/apache/airflow/blob/main/airflow/providers/google/CHANGELOG.rst

Comment (subkanthi) [reply airflow/providers/slack/operators/slack.py]: @mik-laj  I dont think it was working before, the file parameter. I thought I fixed the file-content in an earlier PR, but the passing file was also broken.

Comment (subkanthi) [reply airflow/providers/slack/operators/slack.py]: please let me know , I can add to CHANGELOG, but as I mentioned before, for files, we need to open the file and pass the bufferedreader to Slack API, it was not done before

**Comment by potiuk** [comment]
Some related slack tests are failing. Can you please rebase and fix those ?

**Comment by subkanthi** [comment]
> Some related slack tests are failing. Can you please rebase and fix those ?

Thanks @potiuk , Updated classic case of not digging deeper and assuming that the failure is not related.

**Comment by eladkal** [comment]
@subkanthi it's probably related because the failing tests are of Slack

```
=================================== FAILURES ===================================
  _________ TestSlackAPIFileOperator.test_api_call_params_with_file_args _________
  
  self = <tests.providers.slack.operators.test_slack.TestSlackAPIFileOperator testMethod=test_api_call_params_with_file_args>
  mock_hook = <MagicMock name='SlackHook' id='139893250056768'>
  
      @mock.patch('airflow.providers.slack.operators.slack.SlackHook')
      def test_api_call_params_with_file_args(self, mock_hook):
          test_slack_conn_id = 'test_slack_conn_id'
      
          slack_api_post_operator = SlackAPIFileOperator(
              task_id='slack', slack_conn_id=test_slack_conn_id, file='./test.csv', filetype='csv'
          )
      
  >       slack_api_post_operator.execute()
  
  tests/providers/slack/operators/test_slack.py:215: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  self = <Task(SlackAPIFileOperator): slack>, kwargs = {}
  slack = <MagicMock name='SlackHook()' id='139893328790136'>
  
      def execute(self, **kwargs):
          """
          The SlackAPIOperator calls will not fail even if the call is not unsuccessful.
          It should not prevent a DAG from completing in success
          """
          slack = SlackHook(token=self.token, slack_conn_id=self.slack_conn_id)
      
          # If file content is passed.
          if self.content is not None:
              self.api_params = {
                  'channels': self.channel,
                  'content': self.content,
                  'initial_comment': self.initial_comment,
              }
              slack.call(self.method, data=self.api_params)
          # If file name is passed.
          elif self.file is not None:
              self.api_params = {
                  'channels': self.channel,
                  'filename': self.file,
                  'filetype': self.filetype,
                  'initial_comment': self.initial_comment,
              }
  >           with open(self.file, "rb") as file_name:
  E           FileNotFoundError: [Errno 2] No such file or directory: './test.csv'
  
  airflow/providers/slack/operators/slack.py:242: FileNotFoundError
```

**Comment by subkanthi** [comment]
> @subkanthi it's probably related because the failing tests are of Slack
> 
> ```
> =================================== FAILURES ===================================
>   _________ TestSlackAPIFileOperator.test_api_call_params_with_file_args _________
>   
>   self = <tests.providers.slack.operators.test_slack.TestSlackAPIFileOperator testMethod=test_api_call_params_with_file_args>
>   mock_hook = <MagicMock name='SlackHook' id='139893250056768'>
>   
>       @mock.patch('airflow.providers.slack.operators.slack.SlackHook')
>       def test_api_call_params_with_file_args(self, mock_hook):
>           test_slack_conn_id = 'test_slack_conn_id'
>       
>           slack_api_post_operator = SlackAPIFileOperator(
>               task_id='slack', slack_conn_id=test_slack_conn_id, file='./test.csv', filetype='csv'
>           )
>       
>   >       slack_api_post_operator.execute()
>   
>   tests/providers/slack/operators/test_slack.py:215: 
>   _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   
>   self = <Task(SlackAPIFileOperator): slack>, kwargs = {}
>   slack = <MagicMock name='SlackHook()' id='139893328790136'>
>   
>       def execute(self, **kwargs):
>           """
>           The SlackAPIOperator calls will not fail even if the call is not unsuccessful.
>           It should not prevent a DAG from completing in success
>           """
>           slack = SlackHook(token=self.token, slack_conn_id=self.slack_conn_id)
>       
>           # If file content is passed.
>           if self.content is not None:
>               self.api_params = {
>                   'channels': self.channel,
>                   'content': self.content,
>                   'initial_comment': self.initial_comment,
>               }
>               slack.call(self.method, data=self.api_params)
>           # If file name is passed.
>           elif self.file is not None:
>               self.api_params = {
>                   'channels': self.channel,
>                   'filename': self.file,
>                   'filetype': self.filetype,
>                   'initial_comment': self.initial_comment,
>               }
>   >           with open(self.file, "rb") as file_name:
>   E           FileNotFoundError: [Errno 2] No such file or directory: './test.csv'
>   
>   airflow/providers/slack/operators/slack.py:242: FileNotFoundError
> ```

Yes thanks I noticed that too, cant reproduce it on my local, still checking..

File: airflow/providers/slack/operators/slack.py [hunk 3]
```diff
@@ -195,49 +195,50 @@ class SlackAPIFileOperator(SlackAPIOperator):
...
         filetype: str = None,
         content: str = None,
         **kwargs,
     ) -> None:
         self.method = 'files.upload'
         self.channel = channel
         self.initial_comment = initial_comment
-        self.filename = filename
+        self.file = file
```
Comment (potiuk) [inline airflow/providers/slack/operators/slack.py]: I think it's better to leave that as file_name. I see no point why we should change it to `file`

File: airflow/providers/slack/operators/slack.py [hunk 4]
```diff
@@ -195,49 +195,50 @@ class SlackAPIFileOperator(SlackAPIOperator):
...
-    def execute(self, **kwargs):
-        """
-        The SlackAPIOperator calls will not fail even if the call is not unsuccessful.
-        It should not prevent a DAG from completing in success
-        """
-        if not self.api_params:
-            self.construct_api_call_params()
-        slack = SlackHook(token=self.token, slack_conn_id=self.slack_conn_id)
-        slack.call(self.method, data=self.api_params, files=self.file_params)
+            with open(self.file, "rb") as file_name:
```
Comment (potiuk) [inline airflow/providers/slack/operators/slack.py]: actually the `file_name` here is quite wrong. The `file_name` is actually file handle, not file_name

**Review by potiuk — CHANGES_REQUESTED** [review]
the file/file_name meaning is reversed here (and add backwards_compatibility change) - they should be swapped

File: airflow/providers/slack/operators/slack.py (hunk 4 above)
Comment (subkanthi) [reply airflow/providers/slack/operators/slack.py]: changed to file_handle, thanks

File: airflow/providers/slack/operators/slack.py (hunk 3 above)
Comment (subkanthi) [reply airflow/providers/slack/operators/slack.py]: reverted back to filename, idea was to keep it consistent with the actual slack API expected file.params, but I understand the breakage concern.

**Comment by subkanthi** [comment]
> the file/file_name meaning is reversed here (and add backwards_compatibility change) - they should be swapped

@potiuk can u please check again when u get a chance

**Review by potiuk — APPROVED** [review]
LGTM

File: airflow/providers/slack/operators/slack.py [hunk 5]
```diff
@@ -162,15 +161,16 @@ class SlackAPIFileOperator(SlackAPIOperator):
...
-        slack = SlackAPIFileOperator(
-            task_id="slack_file_upload",
-            dag=dag,
-            slack_conn_id="slack",
-            channel="#general",
-            initial_comment="Hello World!",
-            filename="hello_world.csv",
-            filetype="csv",
-        )
+        with open("test.txt", "rb") as file:
```
Comment (mik-laj) [inline airflow/providers/slack/operators/slack.py:164]: What's changing here? `file` variable looks unused.

**Pull request merged**